theano/tensor/tests/test_opt.py

```diff
@@ -1941,6 +1941,44 @@ def test_local_subtensor_remove_broadcastable_index():
     f2(xn)
 
 
+def test_subtensor_inc_subtensor():
```

Comment:
Can you add calls to `f(...)` with actual numerical values, and then run these tests with `mode=DebugMode`?
I don't think the "complex case" is correct.
Moreover, we would need tests for the case where an `alloc` is returned.